### PR TITLE
Atomic metadata.

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Multiversion metadata for storages without structure locks.

### CODE REVIEW

**Comment by alesapin** [comment]
`01213_alter_rename_column_zookeeper` cannot reproduce locally and don't understand how it's possible.

**Comment by alesapin** [comment]
TODO:
1. Add a test with `ALTER` during long select. (We are absolutely non blocking now)
2. Add comments.
3. Realize what's wrong with `01213_alter_rename_column_zookeeper`

**Comment by alesapin** [comment]
Stress test (memory) — Killed by signal (in clickhouse-server.log) 
FunctionExtractAllGroups in CI fest.

**Comment by alesapin** [comment]
- [ ] Fix clang-tidy.

**Comment by alesapin** [comment]
Race condition in stress test. https://gist.github.com/alesapin/82c0f645d714eb74de255040c1d08fc2

**Comment by alesapin** [comment]
> Race condition in stress test. https://gist.github.com/alesapin/82c0f645d714eb74de255040c1d08fc2

Fix https://github.com/ClickHouse/ClickHouse/pull/11745/commits/292016e04515f5cddf5a540baddab77952e2db83

**Comment by alesapin** [comment]
````020-06-23 12:29:41 00953_moving_functions:                                                 [ FAIL ] 0.15 sec. - result differs with reference:
2020-06-23 12:29:41 --- /usr/share/clickhouse-test/queries/0_stateless/00953_moving_functions.reference	2020-06-23 11:24:32.000000000 +0300
2020-06-23 12:29:41 +++ /tmp/clickhouse-test/0_stateless/00953_moving_functions.stdout	2020-06-23 12:29:41.321314019 +0300
2020-06-23 12:29:41 @@ -37,4 +37,4 @@
2020-06-23 12:29:41  k	groupArrayMovingSum(v)
2020-06-23 12:29:41  String	Array(UInt64)
2020-06-23 12:29:41  a	[0,1,3,6,10]
2020-06-23 12:29:41 -b	[0,1,3,6,10,15,21,28,36,45]
2020-06-23 12:29:41 +b	[5,11,18,26,35,35,36,38,41,45]
```

Flap?

**Comment by alesapin** [comment]
> ```
> 020-06-23 12:29:41 00953_moving_functions:                                                 [ FAIL ] 0.15 sec. - result differs with reference:
> 2020-06-23 12:29:41 --- /usr/share/clickhouse-test/queries/0_stateless/00953_moving_functions.reference	2020-06-23 11:24:32.000000000 +0300
> 2020-06-23 12:29:41 +++ /tmp/clickhouse-test/0_stateless/00953_moving_functions.stdout	2020-06-23 12:29:41.321314019 +0300
> 2020-06-23 12:29:41 @@ -37,4 +37,4 @@
> 2020-06-23 12:29:41  k	groupArrayMovingSum(v)
> 2020-06-23 12:29:41  String	Array(UInt64)
> 2020-06-23 12:29:41  a	[0,1,3,6,10]
> 2020-06-23 12:29:41 -b	[0,1,3,6,10,15,21,28,36,45]
> 2020-06-23 12:29:41 +b	[5,11,18,26,35,35,36,38,41,45]
> ```
> 
> Flap?

Flap

File: src/Storages/System/StorageSystemTables.cpp [hunk 1]
```diff
@@ -359,46 +360,50 @@ class TablesBlockSource : public SourceWithProgress
...
+                    if ((expression_ptr = metadata_snapshot->getSamplingKeyAST()))
                         res_columns[res_index++]->insert(queryToString(expression_ptr));
                     else
                         res_columns[res_index++]->insertDefault();
                 }
 
                 if (columns_mask[src_index++])
                 {
-                    assert(table != nullptr);
+                    assert(metadata_snapshot != nullptr);
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemTables.cpp:406]: Very small inconsistency...

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**